Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp

```diff
@@ -395,82 +397,114 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
     }
+}
+
+const WebsiteDataStoreConfiguration::Directories& WebsiteDataStore::resolvedDirectories()
+{
+    Locker resolveLocker { m_resolveLock };
+    return *m_resolvedDirectories;
+}
+
+void WebsiteDataStore::dispatchResolveDirectories()
+{
+    ASSERT(RunLoop::isMain());
 
-    // Default paths of WebsiteDataStore created with identifier are not under caches or tmp directory,
-    // so we need to explicitly exclude them from backup.
-    if (configuration->identifier()) {
-        Vector<String> allCacheDirectories = {
-            resolvedMediaCacheDirectory()
-            , resolvedNetworkCacheDirectory()
+    m_hasDispatchedResolveDirectories = true;
+    Ref resolveDirectoriesQueue = WorkQueue::create("com.apple.WebKit.WebsiteDataStore.resolveDirectories", WorkQueue::QOS::UserInteractive);
+    resolveDirectoriesQueue->dispatch([this, protectedThis = Ref { *this }, directories = m_configuration->directories().isolatedCopy()]() mutable {
+        resolveDirectories(directories);
+        {
+            Locker resolveLocker { m_resolveLock };
+            m_resolvedDirectories = crossThreadCopy(WTFMove(directories));
+            m_resolveDirectoriesCondition.notifyOne();
+        }
+        didResolveDirectories();
+        RunLoop::main().dispatch([protectedThis = WTFMove(protectedThis)] { });
+    });
+}
+
+void WebsiteDataStore::didResolveDirectories()
+{
+    Vector<String> allCacheDirectories;
+    auto directories = resolvedDirectories();
+    if (m_configuration->identifier()) {
+        // Default paths of WebsiteDataStore created with identifier are not under caches or tmp directory,
+        // so we need to explicitly exclude them from backup.
+        allCacheDirectories = {
+            directories.mediaCacheDirectory
+            , directories.networkCacheDirectory
 #if ENABLE(ARKIT_INLINE_PREVIEW)
-            , resolvedModelElementCacheDirectory()
+            , directories.modelElementCacheDirectory
 #endif
         };
-        protectedQueue()->dispatch([directories = crossThreadCopy(WTFMove(allCacheDirectories))]() {
-            for (auto& directory : directories)
-                FileSystem::setExcludedFromBackup(directory, true);
-        });
     }
 
-    // Clear data of deprecated types asynchronously.
-    if (auto webSQLDirectory = configuration->webSQLDatabaseDirectory(); !webSQLDirectory.isEmpty()) {
-        protectedQueue()->dispatch([webSQLDirectory = webSQLDirectory.isolatedCopy()]() {
+    // Clear data of deprecated types.
+    protectedQueue()->dispatch([webSQLDirectory = crossThreadCopy(directories.webSQLDatabaseDirectory), applicationCacheDirectory = crossThreadCopy(directories.applicationCacheDirectory), applicationCacheFlatFileSubdirectoryName = crossThreadCopy(directories.applicationCacheFlatFileSubdirectoryName), directoriesToExclude = crossThreadCopy(WTFMove(allCacheDirectories))]() {
```

Comment:
If you call `isolatedCopy()` on the Strings you're using to initialize `allCacheDirectories` above (as I suggested) then you could get rid of the `crossThreadCopy()` for `allCacheDirectories` here and just use `WTFMove()`.